File: spacevil_21052019/java_src/Cell.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.Geometry;
import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Core.InterfacePosition;
import com.spvessel.spacevil.Grid;
import com.spvessel.spacevil.Flags.ItemAlignment;

import java.util.List;

public final class Cell extends Geometry implements InterfacePosition {
    private Grid _parent = null;

    /**
     * @return parent grid of the cell
     */
    public Grid getParentGrid() {
        return _parent;
    }

    Cell(Grid grid, int row, int column) {
        _parent = grid;
        _row_index = row;
        _column_index = column;
    }

    // Indecies
    private int _row_index = 0;

    /**
     * @return cell row number
     */
    public int getRow() {
        return _row_index;
    }

    /**
     * @param index cell row number
     */
    public void setRow(int index) {
        _row_index = index;
    }

    private int _column_index = 0;

    /**
     * @return cell column number
     */
    public int getColumn() {
        return _column_index;
    }

    /**
     * @param index cell column number
     */
    public void setColumn(int index) {
        _column_index = index;
    }

    // Position
    private int _x = 0;
    private int _y = 0;

    /**
     * X position of the cell
     */
    public void setX(int x) {
        _x = x;
    }
    public int getX() {
        return _x;
    }

    /**
     * Y position of the cell
     */
    public void setY(int y) {
        _y = y;
    }
    public int getY() {
        return _y;
    }

    private InterfaceBaseItem _item_link = null;

    /**
     * @return cell item
     */
    public InterfaceBaseItem getItem() {
        return _item_link;
    }

    /**
     * Set item into cell
     */
    public void setItem(InterfaceBaseItem item) {
        _item_link = item;
    }

    void updateBehavior() {
        if (getParentGrid() == null || _item_link == null)
            return;

        List<ItemAlignment> alignment = _item_link.getAlignment();

        if (alignment.contains(ItemAlignment.LEFT)) {
            _item_link.setX(getX() + _item_link.getMargin().left);//
        }
        if (alignment.contains(ItemAlignment.RIGHT)) {
            _item_link.setX(getX() + getWidth() - _item_link.getWidth() - _item_link.getMargin().right);//
        }
        if (alignment.contains(ItemAlignment.TOP)) {
            _item_link.setY(getY() + _item_link.getMargin().top);//
        }
        if (alignment.contains(ItemAlignment.BOTTOM)) {
            _item_link.setY(getY() + getHeight() - _item_link.getHeight() - _item_link.getMargin().bottom);//
        }
        if (alignment.contains(ItemAlignment.HCENTER)) {
            _item_link.setX(getX() + (getWidth() - _item_link.getWidth()) / 2 + _item_link.getMargin().left
                    - _item_link.getMargin().right);//
        }
        if (alignment.contains(ItemAlignment.VCENTER)) {
            _item_link.setY(getY() + (getHeight() - _item_link.getHeight()) / 2 + _item_link.getMargin().top
                    - _item_link.getMargin().bottom);//
        }
    }

    void printCellInfo() {
        System.out.println("X: " + _item_link.getX() + "\n" + "Y: " + _item_link.getY() + "\n" + "Row: " + _row_index
                + "\n" + "Column: " + _column_index + "\n" + "Width: " + getWidth() + "\n" + "ItemW: "
                + _item_link.getWidth() + "\n" + "Height: " + getHeight() + "\n" + "ItemH: " + _item_link.getHeight()
                + "\n");
    }

}

File: spacevil_21052019/net_src/Core/Cell.cs
using SpaceVIL.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace SpaceVIL
{
    public sealed class Cell : Geometry, IPosition
    {
        private IGrid _parent = null;

        /// <returns> parent grid of the cell </returns>
        public IGrid GetParentGrid()
        {
            return _parent;
        }
        internal Cell(IGrid grid)
        {
            _parent = grid;
        }
        internal Cell(IGrid grid, int row, int column) : this(grid)
        {
            _row_index = row;
            _column_index = column;
        }
        internal Cell(IGrid grid, int row, int column, IBaseItem item) : this(grid, row, column)
        {
            _item_link = item;
        }

        //Indecies
        private int _row_index = 0;

        /// <returns> cell row number </returns>
        public int GetRow()
        {
            return _row_index;
        }

        /// <param name="index"> cell row number </param>
        public void SetRow(int index)
        {
            _row_index = index;
        }

        private int _column_index = 0;

        /// <returns> cell column number </returns>
        public int GetColumn()
        {
            return _column_index;
        }

        /// <param name="index"> cell column number </param>
        public void SetColumn(int index)
        {
            _column_index = index;
        }

        //Position
        private int _x = 0;
        private int _y = 0;

        /// <summary>
        /// X position of the cell
        /// </summary>
        public void SetX(int x)
        {
            _x = x;
        }
        public int GetX()
        {
            return _x;
        }

        /// <summary>
        /// Y position of the cell
        /// </summary>
        public void SetY(int y)
        {
            _y = y;
        }
        public int GetY()
        {
            return _y;
        }

        private IBaseItem _item_link = null;

        /// <returns> cell item </returns>
        public IBaseItem GetItem()
        {
            return _item_link;
        }

        /// <param name="item"> Set item into cell </param>
        public void SetItem(IBaseItem item)
        {
            _item_link = item;
        }

        internal void UpdateBehavior()
        {
            if (GetParentGrid() == null || _item_link == null)
                return;

            ItemAlignment alignment = _item_link.GetAlignment();

            if (alignment.HasFlag(ItemAlignment.Left))
            {
                _item_link.SetX(GetX() + _item_link.GetMargin().Left);//
            }
            if (alignment.HasFlag(ItemAlignment.Right))
            {
                _item_link.SetX(GetX() + GetWidth() - _item_link.GetWidth() - _item_link.GetMargin().Right);//
            }
            if (alignment.HasFlag(ItemAlignment.Top))
            {
                _item_link.SetY(GetY() + _item_link.GetMargin().Top);//
            }
            if (alignment.HasFlag(ItemAlignment.Bottom))
            {
                _item_link.SetY(GetY() + GetHeight() - _item_link.GetHeight() - _item_link.GetMargin().Bottom);//
            }
            if (alignment.HasFlag(ItemAlignment.HCenter))
            {
                _item_link.SetX(GetX() + (GetWidth() - _item_link.GetWidth()) / 2 + _item_link.GetMargin().Left - _item_link.GetMargin().Right);//
            }
            if (alignment.HasFlag(ItemAlignment.VCenter))
            {
                _item_link.SetY(GetY() + (GetHeight() - _item_link.GetHeight()) / 2 + _item_link.GetMargin().Top - _item_link.GetMargin().Bottom);//
            }
        }

        internal void PrintCellInfo()
        {
            Console.WriteLine(
                "X: " + _item_link.GetX() + "\n" +
                "Y: " + _item_link.GetY() + "\n" +
                "Row: " + _row_index + "\n" +
                "Column: " + _column_index + "\n" +
                "Width: " + GetWidth() + "\n" +
                "ItemW: " + _item_link.GetWidth() + "\n" +
                "Height: " + GetHeight() + "\n" +
                "ItemH: " + _item_link.GetHeight() + "\n" +
                "Item: " + _item_link?.GetItemName() + " type: " + _item_link?.ToString()
                );
        }
    }
}
